transform_inline.ml

```diff
@@ -0,0 +1,312 @@
+  let compose caller callee ret_var arguments : inlining_input =
+    let callee' = callee
+                  |> replace_vars caller
+                  |> replace_labels caller
+    in
+    (* It is important to generate return label and result variable after
+       replacing variables and labels. Otherwise there is a good chance that
+       they will not remain fresh. For ex - if `res` is present in both caller
+       and callee, then `res_1` as a result variable will not remain fresh once
+       `replace_vars` is called. Because `replace_vars` will replace `res` in
+       callee with `res_1` and `res_1` is already used for the result variable.
+       This problem will not happen if we `replace_vars` before generating a
+       fresh `res_var`. *)
+    let ret_lab = "inl"
+                  |> fresh_label callee'.instrs
+                  |> fresh_label caller.instrs
+    in
+    let res_var = "res"
+                  |> fresh (function_vars callee')
+                  |> fresh (function_vars caller)
+    in
+    callee'
+    |> replace_returns res_var ret_lab
+    |> insert_prologue res_var arguments
+    |> insert_epilogue res_var ret_var ret_lab
+  in
+
+  (* Given the caller and callee identifiers, generate a fresh version of caller
+     with the callee inlined.*)
+  let inline_pair (clr : identifier) (cle : identifier) (prog : program) =
+    let caller = lookup_fun prog clr in
+    let caller' = as_inlining_input caller (List.hd caller.body) in
+    let callee = lookup_fun prog cle in
+    let callee' = as_inlining_input callee (List.hd callee.body) in
+    let (pc, ret_var, arguments) = extract_callsite caller'.instrs cle in
+    let callee'' = compose caller' callee' ret_var arguments in
+    let instrs, _ = subst caller'.instrs pc 1 callee''.instrs in
+    add_version caller instrs
+  in
+
+(* This function computes an order for inlining of the entire program.
+   Given a call graph starting from main, it descends as deep as possible into
+   the call chain and stops when it encounters an edge which leads to recursion.
+   Using this depth-first approach, it generates all caller-callee pairs with
+   the depth at which this pair was generated. This depth is used to sort the
+   list of these pairs such that the inlining happens in a bottom up fashion.
+   If a caller-callee pair is already encountered, then it is not included
+   again. This is because once the callee has been inlined in the caller at a
+   given callsite, then the callsite no longer exists. However, multiple
+   callsites with the same target will result in as many caller-callee pairs.
+   Note - The word reduced is used in the general sense of something being
+   simplified. If a function has been analyzed for inlining, its considered
+   reduced. In reality, inlining expands this function. *)
+  let compute_inlining_order (prog : program) : (identifier * identifier) list =
+    let rec inspect_instr ((ord, clr, vis, red, dep) as acc) instr =
```

Comment:
I'm having a hard time following this function. (It doesn't help that it's late and this is the last function I have to review.)

In particular, I can't follow the terse names like `ord`, `clr`, `vis`, `red`, and `dep`. (`acc` is commonly used so it's fine.)

Also, passing a 5-tuple around is kind of awkward; I'd be tempted to just define a new record type. That would help with readability, and also in the places where you only need to change one field.

Finally, I think it's a bit excessive when you have names like `ord`, `ord'`, `ord''`, and `ord'''`.